extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -492,130 +655,58 @@ class LifecycleTask extends BackbeatTask {
-     *   version id marker for versioned buckets
-     * @param {string} [bucketData.details.marker] - next
-     *   continuation token marker for non-versioned buckets
-     * @param {array} bucketLCRules - array of bucket lifecycle rules
-     * @param {array} object - object or object version
+     * @param {object} version - single version from `_getObjectVersions`
+     * @param {object} rules - most applicable rules from `_getApplicableRules`
      * @param {Logger.newRequestLogger} log - logger object
      * @param {function} done - callback(error, data)
      * @return {undefined}
      */
-    _getRules(bucketData, bucketLCRules, object, log, done) {
-        const tagParams = { Bucket: bucketData.target.bucket, Key: object.Key };
-        if (object.VersionId) {
-            tagParams.VersionId = object.VersionId;
-        }
-
-        const req = this.s3target.getObjectTagging(tagParams);
-        attachReqUids(req, log);
-        return req.send((err, tags) => {
-            if (err) {
-                log.error('failed to get tags', {
-                    method: 'LifecycleTask._getRules',
-                    error: err,
-                    bucket: bucketData.target.bucket,
-                    objectKey: object.Key,
-                    objectVersion: object.VersionId,
-                });
-                return done(err);
-            }
-            // tags.TagSet === [{ Key: '', Value: '' }, ...]
-            const filterRules = this._filterRules(bucketLCRules, object, tags);
-
-            // reduce filteredRules to only get earliest dates
-            return done(null, this._getApplicableRules(filterRules));
-        });
-    }
-
-    _compareContents(bucketData, lcRules, contents, log, done) {
-        if (!contents) {
+    _compareVersion(bucketData, version, rules, log, done) {
+        // if version is latest, only expiration action applies
+        if (version.IsLatest) {
+            // TODO: handle unique cases for comparing expiration rule
             return done();
         }
-        return async.eachLimit(contents, CONCURRENCY_DEFAULT, (obj, cb) => {
-            async.waterfall([
-                next => this._getRules(bucketData, lcRules, obj, log, next),
-                (applicableRules, next) => this._compareObject(
-                    bucketData, obj, applicableRules, log, next),
-            ], cb);
-        }, done);
-    }
 
-    _compareVersions(bucketData, lcRules, versions, log, done) {
-        if (!versions) {
-            return done();
+        const staleDate = version.staleDate;
+        if (!staleDate) {
+            // NOTE: this should never happen. Logging here for debug purposes
+            log.error('missing staleDate on the version', {
+                method: 'LifecycleTask._compareVersion',
+                bucket: bucketData.target.bucket,
+                versionId: version.VersionId,
+            });
+            return done(true);
```

Comment:
I think it's best to return an internal error object (possibly with a customized description, although it's more of a sanity check than anything else) rather than a boolean as the error.